### DISCUSSION TITLE
Reset a query back to idle and rerender

### DISCUSSION THREAD

**Discussion by robcaldecott** [TanStack/query#1027] [General]
I am using a `useQuery` with `enabled: false` for a user search, triggering the request using `refetch`. I want to offer a Reset button that will remove the existing query and rerender the form: basically resetting everything back to the initial `idle` state.

I can use `queryCache.invalidateQueries("key")` to invalidate the query but am struggling to find an elegant way to rerender when that happens. When using `useMutation` however, you can call `reset` and this seems to do the trick. However, in this case I am using a `GET` where `useMutation` feels like the wrong fit.

How are other people managing this pattern?

e.g.

```js
const query = useQuery("key", fetchData, { enabled: false }); 
const queryCache = useQueryCache();

{query.isIdle && <MyForm onSubmit={query.refetch} />

{query.isSuccess && <Data data={query.data}>
  <button onClick={() =>{
    queryCache.removeQueries("key");
    // Force rerender with query back to "idle"
}}>Reset</button>
</Data>}
```

**Comment by joeldmyers** [comment]
I think you can pull the `remove` method off of useQuery to do the same?  So change it so that onClick does: 

```
<button onClick={() => {
  query.remove();
}}>Reset</button>
```
https://react-query.tanstack.com/reference/useQuery (remove is down at the bottom).

That _should_ reset isIdle and isSuccess to the pre-fetched status and force a re-render.

**Comment by acontreras89** [comment]
I too would love to have this feature added. I've been using different approaches to achieve this:

* Using `queryClient.resetQueries(queryKey)`
* Using `query.remove()` and `query.refetch()` together

The former is a bit annoying since need to use `useQueryClient`, which ends up bloating the code a little bit.
The later is OK, although I realized these methods change in every (or most) render(s). 
In any case, it would be nice to have this out of the box.

**Comment by TkDodo** [comment]
> The former is a bit annoying since need to use useQueryClient, which ends up bloating the code a little bit.

you can return a `reset` method from a custom hook.

> The later is OK, although I realized these methods change in every (or most) render(s).

the `remove` method returned from `useQuery` is referentially stable between renders. There is no `reset` method returned from `useQuery` 🤔 ?

**Reply by acontreras89** [reply]
Sorry, I meant the `query.refetch()` method. I've updated my previous comment.

**Reply by TkDodo** [reply]
would a `reset` method returned from `useQuery` help?

**Reply by acontreras89** [reply]
Definitely, I'd love to have it out of the box 🙃  even if it is just an alias for `remove` and `refetch`

**Reply by acontreras89** [reply]
Right now I'm using the following hook whenever I have this need:

```js
import { useQuery } from 'react-query'

const useResettableQuery = (...args) => {
  const query = useQuery(...args)

  // NOTE add reset method to the query (shorthand for remove and refetch)
  // NOTE no need to turn this into a callback, since remove and refetch will
  // change in every render
  query.reset = () => {
    query.remove()
    query.refetch()
  }

  return query
}

export default useResettableQuery
```

**Reply by TkDodo** [reply]
`remove` and `refetch` do _not_ change in every render, they are referentially stable. Apart from that, it looks like it gets the job done, however, it's not the same as `queryClient.resetQueries`, especially if you work with `initialData` and `staleTime`. In that case, `reset` would just set it back to the initial data, and if you have a staleTime set, no refetch would happen. So I think what you'd want is:

```
const useResettableQuery = (key, ...args) => {
  const queryClient = useQueryClient()
  const query = useQuery(key, ...args)

  query.reset = () => {
    queryClient.resetQueries(key, { exact: true })
  }

  return query
}
```

this is also very likely how I would implement it in the library, so if you want, you can open a PR :)

**Reply by TkDodo** [reply]
I've opened a draft pull request, can you please give it a try if it behaves as expected?

https://github.com/tannerlinsley/react-query/pull/2637

**Reply by acontreras89** [reply]
Sure, I'll do so first thing tomorrow, thanks!

**Reply by acontreras89** [reply]
@TkDodo I've performed some tests from my app and everything seems to be working just fine. Do you know when can we expect this feature to be released? Do you need any help getting this finished (e.g. updating the docs or anything)?

**Reply by TkDodo** [reply]
docs and tests are missing, didn't get around to it. if you add that, we could merge it I think

**Reply by acontreras89** [reply]
Sure, no prob 🙃  any guidance you can provide would be appreciated, tho (e.g. where to find similar tests)

**Reply by TkDodo** [reply]
I guess the `remove` test might be similar?
https://github.com/tannerlinsley/react-query/blob/bac853fb3c5f4f2c1fb1fa15bc3be9368494ef3d/src/react/tests/useQuery.test.tsx#L949

**Reply by acontreras89** [reply]
@TkDodo I opened [a PR against your fork](https://github.com/TkDodo/react-query/pull/1) with updated docs and tests for the new `reset` function. Let me know if you need anything else.

**Comment by SakarDev** [comment]
> queryClient.resetQueries({ keyName, exact: true });

 as written in their [docs](https://tanstack.com/query/v4/docs/reference/QueryClient/#queryclientresetqueries)
